src/hotspot/share/cds/classPrelinker.cpp

```diff
@@ -0,0 +1,224 @@
+
+bool ClassPrelinker::can_archive_resolved_klass(ConstantPool* cp, int cp_index) {
+  assert(!is_in_archivebuilder_buffer(cp), "sanity");
+  assert(cp->tag_at(cp_index).is_klass(), "must be resolved");
+
+  Klass* resolved_klass = get_resolved_klass_or_null(cp, cp_index);
+  assert(resolved_klass != NULL, "must be");
+
+  return can_archive_resolved_klass(cp->pool_holder(), resolved_klass);
+}
+
+void ClassPrelinker::dumptime_resolve_constants(InstanceKlass* ik, TRAPS) {
+  constantPoolHandle cp(THREAD, ik->constants());
+  if (cp->cache() == NULL || cp->reference_map() == NULL) {
+    // The cache may be NULL if the pool_holder klass fails verification
+    // at dump time due to missing dependencies.
+    return;
+  }
+
+  bool first_time;
+  _processed_classes.put_if_absent(ik, &first_time);
+  if (!first_time) {
+    return;
+  }
+
+  for (int cp_index = 1; cp_index < cp->length(); cp_index++) { // Index 0 is unused
+    switch (cp->tag_at(cp_index).value()) {
+    case JVM_CONSTANT_UnresolvedClass:
+      maybe_resolve_class(cp, cp_index, CHECK);
+      break;
+
+    case JVM_CONSTANT_String:
+      resolve_string(cp, cp_index, CHECK); // may throw OOM when interning strings.
+      break;
+    }
+  }
+}
+
+Klass* ClassPrelinker::find_loaded_class(JavaThread* THREAD, oop class_loader, Symbol* name) {
+  Handle h_loader(THREAD, class_loader);
+  Klass* k = SystemDictionary::find_instance_or_array_klass(THREAD, name,
+                                                            h_loader,
+                                                            Handle());
+  if (k != NULL) {
+    return k;
+  }
+  if (class_loader == SystemDictionary::java_system_loader()) {
+    return find_loaded_class(THREAD, SystemDictionary::java_platform_loader(), name);
+  } else if (class_loader == SystemDictionary::java_platform_loader()) {
+    return find_loaded_class(THREAD, NULL, name);
+  }
+
+  return NULL;
+}
+
+Klass* ClassPrelinker::maybe_resolve_class(constantPoolHandle cp, int cp_index, TRAPS) {
+  assert(!is_in_archivebuilder_buffer(cp()), "sanity");
+  InstanceKlass* cp_holder = cp->pool_holder();
+  if (!cp_holder->is_shared_boot_class() &&
+      !cp_holder->is_shared_platform_class() &&
+      !cp_holder->is_shared_app_class()) {
+    // Don't trust custom loaders, as they may not be well-behaved
+    // when resolving classes.
+    return NULL;
+  }
+
+  CPKlassSlot kslot = cp->klass_slot_at(cp_index);
+  int name_index = kslot.name_index();
+  Symbol* name = cp->symbol_at(name_index);
+  Klass* resolved_klass = find_loaded_class(THREAD, cp_holder->class_loader(), name);
+  if (resolved_klass != NULL && can_archive_resolved_klass(cp_holder, resolved_klass)) {
+    Klass* k = ConstantPool::klass_at_impl(cp, cp_index, CHECK_NULL); // Should fail only with OOM
+    assert(k == resolved_klass, "must be");
+  }
+
+  return resolved_klass;
+}
+
+#if INCLUDE_CDS_JAVA_HEAP
+void ClassPrelinker::resolve_string(constantPoolHandle cp, int cp_index, TRAPS) {
```

Comment:
I prefer to make such methods private, so it can access other private members of the same class.

Reply:
but this doesn't reference other private members of the class. The nice thing about it being a static non-member function is that you can easily see this without looking anywhere else. Also, it can be defined above the caller, which is where it's easy to find.
It's something that varies with developers but I like the style where you read a new file and see little function, little function, etc, leading up to the big function that calls them all.